doc/services/pm/device.rst

```diff
@@ -20,90 +27,174 @@ in power saving mode. This method allows saving power even when the CPU is
 
-Device Power Management Operations
-**********************************
+This type of device power management can be useful when the application is not
+power aware and does not implement runtime device power management. Though,
+:ref:`Device Runtime Power Management <pm-device-runtime-pm>` is the **preferred**
+option for device power management.
 
-Zephyr RTOS power management subsystem provides a control function interface
-to device drivers to indicate power management operations to perform. Each
-device driver defines:
+.. note::
 
-* The device's supported power states.
-* The device's supported transitions between power states.
-* The device's necessary operations to handle the transition between power states.
+    When using this type of device power management, the CPU will only enter
+    a low power state only if no device is in the middle of a hardware
+    transaction that cannot be interrupted.
 
-The following are some examples of operations that the device driver may perform
-in transition between power states:
+.. note::
 
-* Save/Restore device states.
-* Gate/Un-gate clocks.
-* Gate/Un-gate power.
-* Mask/Un-mask interrupts.
+    This type of device power management is disabled when
+    :kconfig:option:`CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE` is set to `y` (that is
+    the default value when :kconfig:option:`CONFIG_PM_DEVICE_RUNTIME` is enabled)
 
-.. note::
+Device Power Management States
+******************************
 
-   When using :ref:`pm-system`, device transitions can be run from the idle thread.
-   As functions in this context cannot block, transitions that intend to use blocking
-   API's **must** check whether they can do so with :c:func:`k_can_yield`.
+The power management subsystem defines device states in
+:c:enum:`pm_device_state`. This type is used to track power states of
+a particular device. It is important to emphasize that, although the
+state is tracked by the subsystem, it is the responsibility of each device driver
+to handle device actions(:c:enum:`pm_device_action`) which change device state.
+
+Each :c:enum:`pm_device_action` have a direct an unambiguous relationship with
+a :c:enum:`pm_device_state`.
+
+.. graphviz::
+   :caption: Device actions x states
+
+    digraph {
+        node [shape=box];
+        rankdir=LR;
+
+        SUSPENDED [label=PM_DEVICE_STATE_SUSPENDED];
+        ACTIVE [label=PM_DEVICE_STATE_ACTIVE];
+        SUSPENDING [label=PM_DEVICE_STATE_SUSPENDING];
+        OFF [label=PM_DEVICE_STATE_OFF];
+
+        SUSPEND [label=PM_DEVICE_ACTION_SUSPEND];
+        RESUME [label=PM_DEVICE_ACTION_RESUME];
+        TURN_OFF [label=PM_DEVICE_ACTION_TURN_OFF];
+        TURN_ON [label=PM_DEVICE_ACTION_TURN_ON];
+        FORCE_SUSPEND [label=PM_DEVICE_ACTION_FORCE_SUSPEND];
+
+        SUSPEND -> SUSPENDING -> SUSPENDED;
+        FORCE_SUSPEND -> SUSPENDING;
+        RESUME -> ACTIVE;
+        TURN_OFF -> OFF;
+        TURN_ON -> ACTIVE;
```

Comment:
So I notice that `TURN_ON` and `TURN_OFF` have never really been properly defined in their transitions. When I implemented them, the intention was that `TURN_OFF` should only ever be called from `STATE_SUSPENDED`, and that `TURN_ON` would return the device to `STATE_SUSPENDED`.

The intention there is to simplify device driver logic, as there is only a single path from `ACTIVE` to `OFF` (via `SUSPENDED`). Otherwise the device needs to query its own state to work out what to do (which we moved away from in transitioning to the action based API).

This gives a graph like
```
digraph {
    node [shape=box];
    rankdir=LR;

    SUSPENDED [label=PM_DEVICE_STATE_SUSPENDED];
    ACTIVE [label=PM_DEVICE_STATE_ACTIVE];
    OFF [label=PM_DEVICE_STATE_OFF];

    OFF -> SUSPENDED ["label"="PM_DEVICE_ACTION_TURN_ON"]
    SUSPENDED -> OFF ["label"="PM_DEVICE_ACTION_TURN_OFF"]

    SUSPENDED -> ACTIVE ["label"="PM_DEVICE_ACTION_RESUME"]
    ACTIVE -> SUSPENDED ["label"="PM_DEVICE_ACTION_SUSPEND"]
}
```

Reply:
Good catch ! I'll update the "TURN_ON" transition. Regarding the suspending, I don't want to
mention device runtime pm here, I will just show these two possible transitions:
1) active -> suspended
2) active -> suspending -> suspended

Reply:
Regarding the `ACTIVE` to `OFF` path, there is nothing blocking the
path from `ACTIVE` to `OFF` without going through suspended.

POWER_DOMAIN (ACTIVE) { DEVICE_A (ACTIVE), DEVICE_B (ACTIVE)};

pm_device_runtime_put(POWER_DOMAIN);

POWER_DOMAIN (supended) --> DEVICE_A (TURN_OFF)
                        |
						--> DEVICE_B (TURN_OFF)

Reply:
I agree that there is nothing enforcing the transition through `SUSPENDED`, but your example is incorrect API usage IMO.
If `DEV_A` and `DEV_B` are active, then they have previously called `pm_device_runtime_get` on the power domain.
Therefore the manual `put` won't depower the domain. Additionally, you have a manual `put` without a corresponding `get`, there is no surprise that things can end up in a weird state.

Reply:
I agree with you, but this is possible. We are documenting the possible transitions here, and right that is possible. I think in guidelines we recommend always calling the child first. But the problem right, is that we have some devices doing runtime PM and others not. So, it is very possible this scenario and it should be documented IMHO.

I'll be super glad to remove it when device runtime will be the only way to do device pm.